File: abelkhan_type_csharp_component_msgpack/component/center/gm_msg_handle.cs
/*
 * gm_msg_handle
 * 2020/6/2
 * qianqians
 */

namespace abelkhan
{
    public class gm_msg_handle
    {
        private abelkhan.gm_center_module gm_center_module;
        private svrmanager svrmng;
        private hubmanager hubmng;
        private gmmanager gmmng;
        private closehandle closeHandle;
        private center_cmd_dispatcher _cmd_dispatcher;

        public gm_msg_handle(abelkhan.modulemng modules, svrmanager svrs, hubmanager hubs, gmmanager gms, closehandle _closeHandle, center_cmd_dispatcher _Dispatcher)
        {
            svrmng = svrs;
            hubmng = hubs;
            gmmng = gms;
            closeHandle = _closeHandle;
            gm_center_module = new abelkhan.gm_center_module(modules);
            _cmd_dispatcher = _Dispatcher;

            gm_center_module.onconfirm_gm += confirm_gm;
            gm_center_module.onclose_clutter += close_clutter;
            gm_center_module.onreload += reload;
            gm_center_module.onreq_cmd += do_cmd;
        }

        private void confirm_gm(string gm_name)
        {
            gmmng.reg_gm(gm_name, gm_center_module.current_ch);
        }

        private void close_clutter(string gmname)
        {
            if (gmmng.check_gm(gmname, gm_center_module.current_ch))
            {
                log.trace(new System.Diagnostics.StackFrame(true), timerservice.Tick, "close_clutter {0}", gmname);

                closeHandle.is_closing = true;
                svrmng.for_each_svr((svrproxy _svrproxy) => {
                    if (_svrproxy.type != "dbproxy")
                    {
                        _svrproxy.close_server();
                    }
                });

                if (hubmng.check_all_hub_closed())
                {
                    svrmng.close_db();
                    closeHandle.is_close = true;
                }
            }
        }

        private void reload(string gmname)
        {
            if (gmmng.check_gm(gmname, gm_center_module.current_ch))
            {
                hubmng.for_each_hub((hubproxy _proxy) => {
                    _proxy.reload();
                });
            }
        }

        private async void do_cmd(string cmd, string param)
        {
            var rsp = (rsp_req_cmd)(gm_center_module.rsp);
            string respStr = await _cmd_dispatcher.Dispatch(cmd, param);
            rsp.rsp(respStr);
        }
    }
}

File: abelkhan_type_component/component/center/gm_msg_handle.ts
import abelkhan = require('../../abelkhan_type/ts/abelkhan');
import center_protcol = require('../../protcol/ts/center');
import svr = require('./svrmanager');
import hub = require('./hubmanager');
import gm = require('./gmmanager');
import close_handle = require('./closehandle');

export class gm_msg_handle{
    private gm_center_module : center_protcol.gm_center_module;
    private svrmng : svr.svrmanager;
    private hubmng : hub.hubmanager;
    private gmmng : gm.gmmanager;
    private close_handle : close_handle.closehandle;
    constructor(modules:abelkhan.modulemng, svrs:svr.svrmanager, hubs:hub.hubmanager, gms:gm.gmmanager, _close_handle:close_handle.closehandle){
        this.svrmng = svrs;
        this.hubmng = hubs;
        this.gmmng = gms;
        this.close_handle = _close_handle;
        this.gm_center_module = new center_protcol.gm_center_module(modules);

        this.gm_center_module.cb_confirm_gm = this.confirm_gm.bind(this);
        this.gm_center_module.cb_close_clutter = this.close_clutter.bind(this);
        this.gm_center_module.cb_reload = this.reload.bind(this);
    }

    public confirm_gm(gm_name:string){
		this.gmmng.reg_gm(gm_name, this.gm_center_module.current_ch);
    }
    
    public close_clutter(gmname:string){
		if (this.gmmng.check_gm (gmname, this.gm_center_module.current_ch)) {
            this.close_handle.is_closing = true;
			this.svrmng.for_each_svr((_svrproxy:svr.svrproxy) => {
                if (_svrproxy.type != "dbproxy") {
                    _svrproxy.close_server();
                }
			});

            if (this.hubmng.check_all_hub_closed()) {
                this.svrmng.close_db();
                this.close_handle.is_close = true;
            }
        }
    }
    
    public reload(gmname:string){
        if (this.gmmng.check_gm(gmname, this.gm_center_module.current_ch)){
            this.hubmng.for_each_hub((_proxy:hub.hubproxy) => {
                _proxy.reload();
            });
        }
    }
}
